R/exercise.R

```diff
@@ -918,21 +918,108 @@ exercise_check_code_is_parsable <- function(exercise) {
     }
   }
 
-  exercise_result(
-    list(
-      message = HTML(
-        i18n_span(
-          "text.unparsable",
-          HTML(i18n_translations()$en$translation$text$unparsable)
-        )
-      ),
-      correct = FALSE,
-      location = "append",
-      type = "error"
-    ),
-    html_output = error_message_html(error$message),
-    error_message = error$message
+  default_message <- i18n_span(
+    "text.unparsable",
+    HTML(i18n_translations()$en$translation$text$unparsable)
   )
+  unicode_message <- exercise_check_unparsable_unicode(exercise)
+
+  feedback <- list(
+    message = HTML(unicode_message %||% default_message),
+    correct = FALSE,
+    location = "append",
+    type = "error"
+  )
+
+  exercise_result_error(error$message, feedback)
+}
+
+exercise_check_unparsable_unicode <- function(exercise) {
+  code <- exercise[["code"]]
+
+  # Early exit if code is made up of all ASCII characters ----------------------
+  if (!grepl("[^\\x00-\\x7F]", code, perl = TRUE)) {
+    return(NULL)
+  }
+
+  # Check if code contains Unicode quotation marks -----------------------------
+  single_quote_pattern <- "[\u2018\u2019\u201A\u201B\u275B\u275C\uFF07]"
+  double_quote_pattern <- "[\u201C-\u201F\u275D\u275E\u301D-\u301F\uFF02]"
+  quote_pattern <- paste(single_quote_pattern, double_quote_pattern, sep = "|")
+
+  if (grepl(quote_pattern, code)) {
+    character <- str_extract(code, quote_pattern)
+    lint <- exercise_highlight_unparsable_unicode(code, quote_pattern)
+
+    # Replace curly single quotes with straight single quotes
+    suggestion <- gsub(single_quote_pattern, "'", code)
+    # Replace all other Unicode quotes with straight double quotes
+    suggestion <- gsub(double_quote_pattern, '"', suggestion)
+    suggestion <- html_code_block(suggestion)
+
+    return(
+      build_unparsable_i18n_span(
+        "unparsablequotes", character, lint, suggestion
+      )
```

Comment:
very minor nit: I'd rather everything on one line if we're not adding argument names

```suggestion
      build_unparsable_i18n_span("unparsablequotes", character, lint, suggestion)
```

otherwise, I'd prefer each argument on a new line and most named (but this looks a little odd, so I think the first is better)

```suggestion
      build_unparsable_i18n_span(
        "unparsablequotes", 
        character = character, 
        lint = lint,
        suggestion = suggestion
      )
```